feat: Create Benchmark Guide for Kong Gateway (#6630)
Kong/docs.konghq.com: app/_src/gateway/production/benchmark.md

You must also ensure that upstream servers are not autoscaled.

### Client maxing out

**Action:** The client must use keep-alive connections.

Sometimes, the clients (such as k6 and Apache JMeter) max themselves out. To tune them, you need to understand the client. Increasing the CPU, threads, and connections on clients results in higher resource utilization and throughput.
The client must also use keep-alive connections. For example, [k6](https://k6.io/docs/using-k6/k6-options/reference/#no-connection-reuse) enables keep-alive by default, and [HTTPClient4](https://hc.apache.org/httpcomponents-client-4.5.x/index.html) implementation in Apache JMeter enables keep-alive by default. Please verify that this is set up appropriately for your test setup.

### Custom plugins

**Action:** Ensure custom plugins aren't interfering with performance.  

Custom plugins can sometimes cause issues with performance. First, you should determine if custom plugins are the source of the performance issues. You can do this by measuring three configuration variations:

> * Keep the necessary bundled plugins enabled and disable all custom plugins<!--should you do a benchmark between this step and the next?-->
> * Configure appropriate custom plugins
> <!--are we missing the third config variation?-->

Author: ```suggestion
1. Measure {{site.base_gateway}}’s performance without enabling any plugins. This provides a baseline for {{site.base_gateway}}’s performance.
1. Enable necessary bundled plugins (plugins that come with the product), and then measure {{site.base_gateway}}’s performance.
1. Next, enable custom plugins (in addition to bundled plugins), and then measure {{site.base_gateway}}’s performance once again.
```

> <!--are we missing the third config variation?-->

Author: Asked Harry these questions on the Confluence page.

> Sometimes, measuring {{site.base_gateway}}'s baseline performance without any plugins enabled can help you find issues outside {{site.base_gateway}}.

Author: ```suggestion
If {{site.base_gateway}}’s baseline performance is poor, then it's likely that either {{site.base_gateway}}’s configuration needs tuning or external factors are affecting it. For external factors, see the other sections in this guide.
A large difference between the performance in the second and and third steps indicates that performance problems could be due to custom plugins.
```

Scale {{site.base_gateway}} out before testing the benchmark to avoid auto-scaling during the benchmark. Monitor the number of {{site.base_gateway}} nodes to ensure new nodes are spawned during the benchmark and existing nodes are not replaced.

### Use multiple cores effectively

**Action:** On most VM setups, set `nginx_worker_processes` to `auto`. On Kubernetes, set `nginx_worker_processes` to one or two less than the worker node CPUs. 

Make sure `nginx_worker_processes` is configured correctly:

* On most VM setups, set this to `auto`. This is the default setting. This ensures that NGINX spawns one worker process for each CPU core, which is desired.
* We recommend setting this explicitly in Kubernetes. Ensure CPU requests and limits for {{site.base_gateway}} match the number of workers configured in {{site.base_gateway}}. For example, if you configure `nginx_worker_processes=4`, you must request 4 CPUs in your pod spec.

>   If you run {{site.base_gateway}} pods on Kubernetes worker nodes with n CPUs, allocate n-2 or n-1 to {{site.base_gateway}}, and configure a worker process count equal to this number. This ensures that any configured daemons and Kubernetes processes, like kubelet, don't contend for resources with {{site.base_gateway}}.  

Reviewer: ```suggestion
  
  If you run {{site.base_gateway}} pods on Kubernetes worker nodes with n CPUs, allocate n-2 or n-1 to {{site.base_gateway}}, and configure a worker process count equal to this number. This ensures that any configured daemons and Kubernetes processes, like kubelet, don't contend for resources with {{site.base_gateway}}.  
```

>   Each additional worker uses additional memory, so you must ensure that {{site.base_gateway}} isn't triggering the Linux Out-of-Memory Killer.

Reviewer: ```suggestion
  
  Each additional worker uses additional memory, so you must ensure that {{site.base_gateway}} isn't triggering the Linux Out-of-Memory Killer.
```

### Resource contention

**Action:** Make sure the client (like Apache JMeter or k6), {{site.base_gateway}}, and upstream servers are on different machines (VM or bare metal) and run on the same local network with low latencies.

> * Ensure that the client (like Apache JMeter or k6), {{site.base_gateway}}, and the upstream servers run on different machines (VM or bare-metal). If these are all running in a Kubernetes cluster, ensure that the pods for these three systems are scheduled on dedicated nodes. Resource contention (of CPU and network usually) between these can lead to sub-optimal performance of any system. 

Reviewer: ```suggestion
* Ensure that the client (like Apache JMeter or k6), {{site.base_gateway}}, and the upstream servers run on different machines (VM or bare-metal). If these are all running in a Kubernetes cluster, ensure that the pods for these three systems are scheduled on dedicated nodes. Resource contention (usually CPU and network) between these can lead to suboptimal performance of any system. 
```

* Ensure the client, {{site.base_gateway}}, and upstream servers run on the same local network with low latencies. If requests between the client and {{site.base_gateway}} or {{site.base_gateway}} and the upstream server traverse the internet, then the results will contain unnecessary noise. 

### Upstream servers maxing out

**Action:** Verify that the upstream server isn't maxing out.

You can verify that the upstream server isn't maxing out by checking the CPU and memory usage of the upstream server.
If you deploy additional {{site.base_gateway}} nodes and the throughput or error rate remains the same, the upstream server or a system other than {{site.base_gateway}} is likely the bottleneck.

> You must also ensure that upstream servers are not autoscaled.

Reviewer: ```suggestion

You must also ensure that upstream servers are not autoscaled.
```

### Client maxing out

**Action:** The client must use keep-alive connections.

Sometimes, the clients (such as k6 and Apache JMeter) max themselves out. To tune them, you need to understand the client. Increasing the CPU, threads, and connections on clients results in higher resource utilization and throughput.

> The client must also use keep-alive connections. For example, [k6](https://k6.io/docs/using-k6/k6-options/reference/#no-connection-reuse) enables keep-alive by default, and [HTTPClient4](https://hc.apache.org/httpcomponents-client-4.5.x/index.html) implementation in Apache JMeter enables keep-alive by default. Please verify that this is set up appropriately for your test setup.

Reviewer: ```suggestion

The client must also use keep-alive connections. For example, [k6](https://k6.io/docs/using-k6/k6-options/reference/#no-connection-reuse) and the [HTTPClient4](https://hc.apache.org/httpcomponents-client-4.5.x/index.html) implementation in Apache JMeter both enable keep-alive by default. 
Verify that this is set up appropriately for your test setup.
```

### Custom plugins

**Action:** Ensure custom plugins aren't interfering with performance.  

Custom plugins can sometimes cause issues with performance. First, you should determine if custom plugins are the source of the performance issues. You can do this by measuring three configuration variations:

1. Measure {{site.base_gateway}}’s performance without enabling any plugins. This provides a baseline for {{site.base_gateway}}’s performance.
1. Enable necessary bundled plugins (plugins that come with the product), and then measure {{site.base_gateway}}’s performance.
1. Next, enable custom plugins (in addition to bundled plugins), and then measure {{site.base_gateway}}’s performance once again.

If {{site.base_gateway}}’s baseline performance is poor, then it's likely that either {{site.base_gateway}}’s configuration needs tuning or external factors are affecting it. For external factors, see the other sections in this guide.

> A large difference between the performance in the second and and third steps indicates that performance problems could be due to custom plugins.

Author: ```suggestion
A large difference between the performance in the second and third steps indicates that performance problems could be due to custom plugins.
```

While AWS is mentioned in the following, the same recommendations apply to most cloud providers:

* Ensure that you are not using burstable instances, like T type instances, in AWS. In this case, the CPU available to applications is variable, which leads to noise in the stats. For more information, see the [Burstable performance instances](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/burstable-performance-instances.html) AWS documentation.
* Ensure you are not hitting bandwidth limits, TCP connections per unit time limits, or Packet Per Second (PPS) limits. For more information, see the [Amazon EC2 instance network bandwidth](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-network-bandwidth.html) AWS documentation.

### Configuration changes during benchmark tests

**Action:** Don't change the {{site.base_gateway}} configuration during a benchmark test.

If you change the configuration during a test, {{site.base_gateway}}'s tail latencies can increase sharply. Avoid doing this unless you are measuring {{site.base_gateway}}'s performance under a configuration change.

### Large request and response bodies

> **Action:** Keep request bodies below eight KB and response bodies below 32 KB.

Reviewer: ```suggestion
**Action:** Keep request bodies below 8 KB and response bodies below 32 KB.
```
I know we usually spell out a number when it's ten or lower, but I think we should revisit this and set some guidelines on when to use what. When it's a measurement (like KB in this instance), it makes more sense to use the numeral. Additionally, consistency/parallelism is important for clarity, and it's awkward to read "eight KB and 32 KB". 

Interestingly, the [Google style guide](https://developers.google.com/style/numbers#numbers-as-numerals) lists a fair number of exceptions here, including units, and the Mailchimp style guide tells writers to [use the numeral in most situations](https://styleguide.mailchimp.com/grammar-and-mechanics/#header-3-numbers).

> Most benchmarking setups generally consist of an HTTP request with a small HTTP body and a corresponding HTTP response with a JSON or HTML response body. A request body of less than eight KB and a response body of less than 32 KB is considered small. If your request or response bodies are larger, {{site.base_gateway}} will buffer the request and response using the disk, which significantly impacts {{site.base_gateway}}'s performance.

Reviewer: ```suggestion
Most benchmarking setups generally consist of an HTTP request with a small HTTP body and a corresponding HTTP response with a JSON or HTML response body. 
A request body of less than 8 KB and a response body of less than 32 KB is considered small. If your request or response bodies are larger, {{site.base_gateway}} will buffer the request and response using the disk, which significantly impacts {{site.base_gateway}}'s performance.
```

The following `kong.conf` file examples contain all the recommended parameters from the previous section:

### Configuration values
```bash
# change this for Kubernetes setup to a number matching the CPU limit. Suggested value is 4 or 8.
nginx_worker_processes=auto

upstream_keepalive_max_requests=100000
nginx_http_keepalive_requests=100000

proxy_access_log=off

dns_stale_ttl=3600
```

### Environment variable format

```bash
# Change this for Kubernetes setup to a number matching CPU limit. We recommend four or eight.
KONG_NGINX_WORKER_PROCESSES="auto"
KONG_UPSTREAM_KEEPALIVE_MAX_REQUESTS="100000"
KONG_NGINX_HTTP_KEEPALIVE_REQUESTS="100000"

KONG_PROXY_ACCESS_LOG="off"

KONG_DNS_STALE_TTL="3600"
```

### Configuration for Helm chart values.yaml

```bash
# The value of four for worker process is a suggested value. You can use eight as well.
# Allocate the same amount of CPU and appropriate memory to avoid OOM killer.
env:
  nginx_worker_processes: "4"
  upstream_keepalive_max_requests: "100000"
  nginx_http_keepalive_requests: "100000"
  proxy_access_log: "off"
  dns_stale_ttl: "3600"

resources:
  requests:
    cpu: 4
    memory: ""
  limits:
    cpu: 4
    memory: ""

> ```

Reviewer: Let's move these into tabs instead of headings: kong.conf, environment variables, and Helm chart values
